### PR TITLE
Login

### PR DESCRIPTION
**Pull request by Prateek0129**

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<div class="container">
```
Comment (upexcel) [inline src/app/app.component.html]: remove html from here. All link will be in login/registration page

File: src/app/app.component.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'app-root',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.css']
+})
+export class AppComponent {
+  title = 'app';
```
Comment (upexcel) [inline src/app/app.component.ts]: remove title. you didn't used it..

File: src/app/login/login.component.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    <label for="pwdbox">
+      Password:
+    </label>
+    <input type="password" autocomplete="off" id="pwdbox" class="form-control" formControlName="password" placeholder="Password" required>
+    <div *ngIf="loginForm.controls['password'].errors && !loginForm.controls['password'].pristine" class="error-msg">
+      <div [hidden]="!loginForm.controls['password'].errors.required" class="alert alert-danger alert-override"><strong>Password</strong> is required.</div>
+    </div>
+  </div>
+  <div>
+    <button type="submit" class="btn btn-danger btn-block btn-alter" [ngClass] = "{'color': !loginForm.invalid }" [disabled]="loginForm.invalid">Login</button>
```
Comment (upexcel) [inline src/app/login/login.component.html]: add a spinner.

File: src/app/login/login.component.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  constructor() { }
+    
+    loginForm: FormGroup;
+  
+    ngOnInit() {
+    this.createForm();
+  }
+  
+  private createForm() {
```
Comment (upexcel) [inline src/app/login/login.component.ts]: why use private ?

File: src/app/register/register.component.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    <div class="form-inside">
+        <label for="cnfpwdbox">
+            Confirm Password:
+        </label>
+        <input type="password" autocomplete="off" id="cnfpwdbox" class="form-control" formControlName="confirmPassword" placeholder="Confirm Password" required>
+        <div *ngIf="registerForm.controls.confirmPassword.errors?.MatchPassword" class="alert alert-danger alert-override"><strong>Password</strong> Do Not Match</div>
+    </div>
+    <div>
+        <button type="submit" class="btn btn-danger btn-md btn-block btn-alter" [ngClass] = "{'color': !registerForm.invalid }" [disabled]="registerForm.invalid">Register</button>
```
Comment (upexcel) [inline src/app/register/register.component.html]: add a spinner

File: src/app/login/login.component.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    this.loginForm = new FormGroup({
+      email: new FormControl('', [Validators.required,patternValidator(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/)]),
+      password: new FormControl('',Validators.required),
```
Comment (upexcel) [inline src/app/login/login.component.ts]: add validation password of min 4 max 8 length and a special character.

File: src/app/register/register.component.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+  constructor( ) { }
+
+  ngOnInit() {
+    this.createForm();
+  }
+  private createForm() {
+    this.registerForm = new FormGroup({
+      email:            new FormControl('',[Validators.required,patternValidator(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/)]),
+      password:         new FormControl('',[Validators.required]),
```
Comment (upexcel) [inline src/app/register/register.component.ts]: add validation password of min 4 max 8 length and a special character.

File: src/app/app.component.html (hunk 1 above)
Comment (Prateek0129) [reply src/app/app.component.html]: done sir.

File: src/app/login/login.component.ts (hunk 6 above)
Comment (Prateek0129) [reply src/app/login/login.component.ts]: done sir.

File: src/app/register/register.component.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      email:            new FormControl('',[Validators.required,patternValidator(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/)]),
+      password:         new FormControl('',[Validators.required,passwordPattern(/^(?=^.{4,8}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&amp;*()_+}{&quot;:;'?/&gt;.&lt;,])(?!.*\s).*$/)]),
+      confirmPassword:  new FormControl('',[Validators.required]),
+      role:             new FormControl('',Validators.required),
+    }, PasswordValidation.MatchPassword);
+  }
+
+  onRegister(post) {
+    this.spin = true
+    this.loginService.onRegister(post).then(()=>{
```
Comment (upexcel) [inline src/app/register/register.component.ts]: create register service.

File: src/app/login.service.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { Injectable } from '@angular/core';
+import { HttpParams, HttpClient } from '@angular/common/http';
+import { promise } from 'selenium-webdriver';
+
+@Injectable()
+export class LoginService {
+login;
```
Comment (upexcel) [inline src/app/login.service.ts]: remove unwanted variables

File: src/app/login.service.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+register;
+  constructor(private http: HttpClient) { }
+  onLogin(post) {
+    return new Promise((resolve,reject)=>{
+      this.http.get(' https://secure-refuge-14993.herokuapp.com/login', {
```
Comment (upexcel) [inline src/app/login.service.ts]: create a common http service that handle http tasks

File: src/app/http.service.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import { promise } from 'selenium-webdriver';
+import { HttpErrorResponse } from '@angular/common/http/src/response';
+
+@Injectable()
+export class HttpService {
+
+  constructor(private http: HttpClient) { }
+  onLogin(post) {
+    return new Promise((resolve, reject) => {
+      this.http.get(' https://secure-refuge-14993.herokuapp.com/login', {
```
Comment (upexcel) [inline src/app/http.service.ts]: create environment variable for apiUrl.

**Pull request merged**